Probable Race Condition in Block Access Controller

Summary: When compiled for release there appears to be an infrequent race condition that will cause the BlockAccessControllerTest to fail the assertion on line 116 which ensures that WaitUntilHot will guarantee all locks are released before making a block hot.

### Test results
```
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from BlockAccessControllerTest
[ RUN      ] BlockAccessControllerTest.SingleReader
[       OK ] BlockAccessControllerTest.SingleReader (0 ms)
[ RUN      ] BlockAccessControllerTest.WriterPreemption
[       OK ] BlockAccessControllerTest.WriterPreemption (0 ms)
[ RUN      ] BlockAccessControllerTest.FreezingBlocksBoth
[       OK ] BlockAccessControllerTest.FreezingBlocksBoth (11 ms)
[ RUN      ] BlockAccessControllerTest.ReaderWriter
[       OK ] BlockAccessControllerTest.ReaderWriter (11 ms)
[ RUN      ] BlockAccessControllerTest.MultipleReaders
../test/storage/block_access_controller_test.cpp:116: Failure
Value of: REACHED(reader1_released) && REACHED(reader1_released)
  Actual: false
Expected: true
[  FAILED  ] BlockAccessControllerTest.MultipleReaders (17 ms)
[----------] 5 tests from BlockAccessControllerTest (39 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (39 ms total)
[  PASSED  ] 4 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] BlockAccessControllerTest.MultipleReaders
```

### Steps to reproduce:
```bash
cmake -DCMAKE_BUILD_TYPE=release -DTERRIER_USE_ASAN=off -DBUILD_WARNING_LEVEL=production
make -j
release/block_access_controller_test --gtest_repeat=-1 --gtest_throw_on_failure
```

### Notes:
Building in release mode results in the following warnings:
> ../src/include/storage/block_access_controller.h: In member function ‘std::pair<terrier::storage::BlockState, unsigned int> terrier::storage::BlockAccessController::AtomicallyLoadMembers()’:
> ../src/include/storage/block_access_controller.h:128:76: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
>      return *reinterpret_cast<std::pair<BlockState, uint32_t> *>(&curr_value);
>                                                                             ^
> ../src/include/storage/block_access_controller.h:128:76: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
> ../src/include/storage/block_access_controller.h: In member function ‘bool terrier::storage::BlockAccessController::UpdateAtomically(const std::pair<terrier::storage::BlockState, unsigned int>&, std::pair<terrier::storage::BlockState, unsigned int>)’:
> ../src/include/storage/block_access_controller.h:134:63: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
>          expected_bytes, *reinterpret_cast<uint64_t *>(&desired));
>                                                                ^

### Initial suspicions:
We are using a struct of two `uin32_t`s to pass the 64-bit type value, which means the values are not guaranteed to be placed 64-bit aligned.  This could cause us to lose the atomicity we are expecting in the code.

--- a/src/include/storage/block_access_controller.h
+++ b/src/include/storage/block_access_controller.h
@@ -44,6 +44,9 @@ enum class BlockState : uint32_t {
  * reading transactionally if the block is not frozen.
  */
 class BlockAccessController {
+  // We do some reinterpret_casting between uint64_t and the std::pair below, so we want to assert the object size.
+  static_assert(sizeof(std::pair<BlockState, uint32_t>) == sizeof(uint64_t));
+
  public:
   /**
    * Initialize the block access controller.
@@ -125,13 +128,14 @@ class BlockAccessController {
 
   std::pair<BlockState, uint32_t> AtomicallyLoadMembers() {
     uint64_t curr_value = reinterpret_cast<std::atomic<uint64_t> *>(bytes_)->load();
-    return *reinterpret_cast<std::pair<BlockState, uint32_t> *>(&curr_value);
+    const auto *const curr_state = reinterpret_cast<const std::pair<BlockState, uint32_t> *const>(&curr_value);
+    return *curr_state;
   }
 
   bool UpdateAtomically(const std::pair<BlockState, uint32_t> &expected, std::pair<BlockState, uint32_t> desired) {
-    uint64_t expected_bytes = *reinterpret_cast<const uint64_t *>(&expected);
-    return reinterpret_cast<std::atomic<uint64_t> *>(bytes_)->compare_exchange_strong(
-        expected_bytes, *reinterpret_cast<uint64_t *>(&desired));
+    auto expected_bytes = *reinterpret_cast<const uint64_t *const>(&expected);
+    const auto *const desired_bytes = reinterpret_cast<const uint64_t *const>(&desired);
+    return reinterpret_cast<std::atomic<uint64_t> *>(bytes_)->compare_exchange_strong(expected_bytes, *desired_bytes);
   }
 };
 }  // namespace terrier::storage

--- a/test/storage/block_access_controller_test.cpp
+++ b/test/storage/block_access_controller_test.cpp
@@ -37,7 +37,7 @@ TEST(BlockAccessControllerTest, WriterPreemption) {
 // Tests that the freezing flag blocks both readers and writers
 // NOLINTNEXTLINE
 TEST(BlockAccessControllerTest, FreezingBlocksBoth) {
-  const uint32_t iteration = 100;
+  const uint32_t iteration = 10000;
   for (uint32_t i = 0; i < iteration; i++) {
     storage::BlockAccessController tested;
     tested.Initialize();
@@ -60,27 +60,27 @@ TEST(BlockAccessControllerTest, FreezingBlocksBoth) {
 // Tests that writer blocks future readers but waits for reader exit
 // NOLINTNEXTLINE
 TEST(BlockAccessControllerTest, ReaderWriter) {
-  const uint32_t iteration = 100;
+  const uint32_t iteration = 10000;
   for (uint32_t i = 0; i < iteration; i++) {
     storage::BlockAccessController tested;
     tested.Initialize();
     tested.GetBlockState()->store(storage::BlockState::FROZEN);
     DECLARE_PROGRAM_POINT(read_acquired)
-    DECLARE_PROGRAM_POINT(read_released)
+    DECLARE_PROGRAM_POINT(read_releasing)
     std::thread reader([&] {
       EXPECT_TRUE(tested.TryAcquireInPlaceRead());
       PROGRAM_POINT(read_acquired)
       // Block the reader until the write has executed
       WAIT_UNTIL(tested.GetBlockState()->load() == storage::BlockState::HOT)
       EXPECT_FALSE(tested.TryAcquireInPlaceRead());
-      PROGRAM_POINT(read_released)
+      PROGRAM_POINT(read_releasing)
       tested.ReleaseInPlaceRead();
     });
 
     WAIT_UNTIL(read_acquired)
     tested.WaitUntilHot();
     // The thread should only be unblocked when the reader has released the lock
-    EXPECT_TRUE(read_released);
+    EXPECT_TRUE(read_releasing);
     EXPECT_EQ(tested.GetBlockState()->load(), storage::BlockState::HOT);
     reader.join();
   }
@@ -89,31 +89,31 @@ TEST(BlockAccessControllerTest, ReaderWriter) {
 // Tests that writer waits for all readers to exit
 // NOLINTNEXTLINE
 TEST(BlockAccessControllerTest, MultipleReaders) {
-  const uint32_t iteration = 100;
+  const uint32_t iteration = 10000;
   for (uint32_t i = 0; i < iteration; i++) {
     storage::BlockAccessController tested;
     tested.Initialize();
     tested.GetBlockState()->store(storage::BlockState::FROZEN);
     DECLARE_PROGRAM_POINT(reader1_acquired)
-    DECLARE_PROGRAM_POINT(reader1_released)
+    DECLARE_PROGRAM_POINT(reader1_releasing)
     DECLARE_PROGRAM_POINT(reader2_acquired)
-    DECLARE_PROGRAM_POINT(reader2_released)
+    DECLARE_PROGRAM_POINT(reader2_releasing)
     std::thread reader1([&] {
       EXPECT_TRUE(tested.TryAcquireInPlaceRead());
       PROGRAM_POINT(reader1_acquired)
+      PROGRAM_POINT(reader1_releasing)
       tested.ReleaseInPlaceRead();
-      PROGRAM_POINT(reader1_released)
     });
 
     std::thread reader2([&] {
       EXPECT_TRUE(tested.TryAcquireInPlaceRead());
       PROGRAM_POINT(reader2_acquired)
+      PROGRAM_POINT(reader2_releasing)
       tested.ReleaseInPlaceRead();
-      PROGRAM_POINT(reader2_released)
     });
     WAIT_UNTIL(reader1_acquired && reader2_acquired)
     tested.WaitUntilHot();
-    EXPECT_TRUE(REACHED(reader1_released) && REACHED(reader1_released));
+    EXPECT_TRUE(REACHED(reader1_releasing) && REACHED(reader1_releasing));
     EXPECT_EQ(tested.GetBlockState()->load(), storage::BlockState::HOT);
     reader1.join();
     reader2.join();

--- a/test/storage/block_access_controller_test.cpp
+++ b/test/storage/block_access_controller_test.cpp
@@ -113,7 +113,7 @@ TEST(BlockAccessControllerTest, MultipleReaders) {
     });
     WAIT_UNTIL(reader1_acquired && reader2_acquired)
     tested.WaitUntilHot();
-    EXPECT_TRUE(REACHED(reader1_releasing) && REACHED(reader1_releasing));
+    EXPECT_TRUE(REACHED(reader1_releasing) && REACHED(reader2_releasing));
     EXPECT_EQ(tested.GetBlockState()->load(), storage::BlockState::HOT);
     reader1.join();
     reader2.join();

